espnet2/samplers/batch_bin_sampler.py

```diff
@@ -0,0 +1,153 @@
+                        "If padding=True, the feature dimension must be unified: {s}",
+                    )
+            # If padding case, the feat-dim must be same over whole corpus,
+            # therefore the first sample is referred
+            feat_dims = [np.prod(d[keys[0]][1:]) for d in utt2shapes]
+        else:
+            feat_dims = None
+
+        # Decide batch-sizes
+        start = 0
+        batch_sizes = []
+        bs = 1
+        while True:
+            # shape: (Length, dim1, dim2, ...)
+            if padding:
+                max_lengths = [
+                    max(d[keys[i]][0] for i in range(start, start + bs))
+                    for d in utt2shapes
+                ]
+                bins = sum(bs * l * d for l, d in zip(max_lengths, feat_dims))
+            else:
+                bins = sum(
+                    np.prod(d[keys[i]])
+                    for i in range(start, start + bs)
+                    for d in utt2shapes
+                )
+
+            if bins > batch_bins and bs >= min_batch_size:
+                batch_sizes.append(bs)
+                start += bs
+                bs = 1
+            else:
+                bs += 1
+            if start >= len(keys):
+                break
+
+            if start + bs > len(keys):
+                if not self.drop_last or len(batch_sizes) == 0:
+                    batch_sizes.append(len(keys) - start)
+                break
+
+        if len(batch_sizes) == 0:
+            # Maybe we can't reach here
+            raise RuntimeError("0 batches")
+
+        # If the last batch-size is smaller than minimum batch_size,
+        # the samples are redistributed to the other mini-batches
+        if len(batch_sizes) > 1 and batch_sizes[-1] < min_batch_size:
+            for i in range(batch_sizes.pop(-1)):
+                batch_sizes[-(i % len(batch_sizes)) - 2] += 1
+
+        if not self.drop_last:
+            # Bug check
+            assert sum(batch_sizes) == len(keys), f"{sum(batch_sizes)} != {len(keys)}"
+
+        # Set mini-batch
+        self.batch_list = []
+        start = 0
+        for bs in batch_sizes:
+            assert len(keys) >= start + bs, "Bug"
+            minibatch_keys = keys[start : start + bs]
+            start += bs
+            if sort_in_batch == "descending":
+                minibatch_keys.reverse()
+            elif sort_in_batch == "ascending":
+                # Key are already sorted in ascending
+                pass
+            else:
+                raise ValueError(
+                    f"sort_in_batch must be ascending or descending: {sort_in_batch}"
+                )
+            self.batch_list.append(tuple(minibatch_keys))
+
+        if sort_batch == "ascending":
+            pass
+        elif sort_batch == "descending":
+            self.batch_list.reverse()
+        else:
+            raise ValueError(
+                f"sort_batch must be ascending or descending: {sort_batch}"
```

Comment:
It would be better If this validation were at the beginning of this func

Reply:
applied.